File: sdk-rust/src/models/account.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AccountOverview {
    pub perp_equity_balance: f64,
    pub unrealized_pnl: f64,
    pub unrealized_funding_cost: f64,
    pub cross_margin_ratio: f64,
    pub maintenance_margin: f64,
    pub cross_account_leverage_ratio: Option<f64>,
    pub volume: Option<f64>,
    pub net_deposits: Option<f64>,
    pub all_time_return: Option<f64>,
    pub pnl_90d: Option<f64>,
    pub sharpe_ratio: Option<f64>,
    pub max_drawdown: Option<f64>,
    pub weekly_win_rate_12w: Option<f64>,
    pub average_cash_position: Option<f64>,
    pub average_leverage: Option<f64>,
    pub cross_account_position: f64,
    pub total_margin: f64,
    pub usdc_cross_withdrawable_balance: f64,
    pub usdc_isolated_withdrawable_balance: f64,
    pub realized_pnl: Option<f64>,
    pub liquidation_fees_paid: Option<f64>,
    pub liquidation_losses: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserSubaccount {
    pub subaccount_address: String,
    pub primary_account_address: String,
    pub is_primary: bool,
    pub custom_label: Option<String>,
    pub is_active: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Delegation {
    pub delegated_account: String,
    pub permission_type: String,
    pub expiration_time_s: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserFundHistoryItem {
    pub amount: f64,
    pub is_deposit: bool,
    pub transaction_unix_ms: i64,
    pub transaction_version: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserFundingHistoryItem {
    pub market: String,
    pub funding_rate_bps: f64,
    pub is_funding_positive: bool,
    pub funding_amount: f64,
    pub position_size: f64,
    pub transaction_unix_ms: i64,
    pub transaction_version: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LeaderboardItem {
    pub rank: i64,
    pub account: String,
    pub account_value: f64,
    pub realized_pnl: f64,
    pub roi: f64,
    pub volume: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Leaderboard {
    pub items: Vec<LeaderboardItem>,
    pub total_count: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PortfolioChartData {
    pub timestamp: i64,
    pub value: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserNotification {
    pub id: String,
    #[serde(rename = "type")]
    pub notification_type: String,
    pub message: String,
    pub timestamp: i64,
    pub read: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserTradeHistoryItem {
    pub account: String,
    pub market: String,
    pub action: String,
    pub size: f64,
    pub price: f64,
    pub is_profit: bool,
    pub realized_pnl_amount: f64,
    pub is_funding_positive: bool,
    pub realized_funding_amount: f64,
    pub is_rebate: bool,
    pub fee_amount: f64,
    pub transaction_unix_ms: i64,
    pub transaction_version: i64,
}

File: sdk-python/decibel_sdk/models/account.py
"""Account, subaccount, delegation, and analytics models."""

from __future__ import annotations

from pydantic import BaseModel, Field


class AccountOverview(BaseModel):
    perp_equity_balance: float
    unrealized_pnl: float
    unrealized_funding_cost: float
    cross_margin_ratio: float
    maintenance_margin: float
    cross_account_leverage_ratio: float | None = None
    volume: float | None = None
    net_deposits: float | None = None
    all_time_return: float | None = None
    pnl_90d: float | None = None
    sharpe_ratio: float | None = None
    max_drawdown: float | None = None
    weekly_win_rate_12w: float | None = None
    average_cash_position: float | None = None
    average_leverage: float | None = None
    cross_account_position: float
    total_margin: float
    usdc_cross_withdrawable_balance: float
    usdc_isolated_withdrawable_balance: float
    realized_pnl: float | None = None
    liquidation_fees_paid: float | None = None
    liquidation_losses: float | None = None


class UserSubaccount(BaseModel):
    subaccount_address: str
    primary_account_address: str
    is_primary: bool
    custom_label: str | None = None
    is_active: bool | None = None


class Delegation(BaseModel):
    delegated_account: str
    permission_type: str
    expiration_time_s: int | None = None


class UserFundHistoryItem(BaseModel):
    amount: float
    is_deposit: bool
    transaction_unix_ms: int
    transaction_version: int


class UserFundingHistoryItem(BaseModel):
    market: str
    funding_rate_bps: float
    is_funding_positive: bool
    funding_amount: float
    position_size: float
    transaction_unix_ms: int
    transaction_version: int


class LeaderboardItem(BaseModel):
    rank: int
    account: str
    account_value: float
    realized_pnl: float
    roi: float
    volume: float


class Leaderboard(BaseModel):
    items: list[LeaderboardItem]
    total_count: int


class PortfolioChartData(BaseModel):
    timestamp: int
    value: float


class UserNotification(BaseModel):
    id: str
    notification_type: str = Field(alias="type")
    message: str
    timestamp: int
    read: bool

    model_config = {"populate_by_name": True}


class UserTradeHistoryItem(BaseModel):
    account: str
    market: str
    action: str
    size: float
    price: float
    is_profit: bool
    realized_pnl_amount: float
    is_funding_positive: bool
    realized_funding_amount: float
    is_rebate: bool
    fee_amount: float
    transaction_unix_ms: int
    transaction_version: int
